### PR TITLE
Refactor ValueTypeNone into ValueTypeNull and ValueTypeNotGiven

### PR DESCRIPTION
**Pull request by fgalan**
Refactor pending related with #2998 

This refactor splits former ValueTypeNone into ValueTypeNull and ValueTypeNotGiven thus removing the ambiguity between "value field exists and asigned as JSON nul" and "no value field".

As a consecuence, the mongoBackend logic is now a bit simpler. Some cryptic and superfluous methods has been removed.

@kzangeli

### CODE REVIEW

File: src/lib/common/macroSubstitute.cpp [hunk 1]
```diff
@@ -57,10 +57,15 @@ static void attributeValue(std::string* valueP, const std::vector<ContextAttribu
...
       *valueP = (vec[ix]->boolValue == true)? "true" : "false";
     }
-    else if (vec[ix]->valueType == orion::ValueTypeNone)
+    else if (vec[ix]->valueType == orion::ValueTypeNull)
     {
       *valueP = "null";
     }
+    else if (vec[ix]->valueType == orion::ValueTypeNotGiven)
+    {
+      LM_E(("Runtime Error (ValueTpeNotGiven for attribute)"));
```
Comment (kzangeli) [inline src/lib/common/macroSubstitute.cpp]: Typo.
However, why not write out as text?  (Value not given for attribute)

File: src/lib/jsonParse/jsonAppendContextElementRequest.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ static std::string contextAttribute(const std::string& path, const std::string&
 {
   LM_T(LmtParse, ("Got an attribute"));
   reqData->acer.attributeP = new ContextAttribute();
-  reqData->acer.attributeP->valueType = orion::ValueTypeNone;
+  reqData->acer.attributeP->valueType = orion::ValueTypeNotGiven;
```
Comment (kzangeli) [inline src/lib/jsonParse/jsonAppendContextElementRequest.cpp]: Perhaps we should put `valueType = orion::ValueTypeNotGiven` in the constructor(s) of `ContextAttribute`?

File: src/lib/jsonParse/jsonUpdateContextResponse.cpp [hunk 3]
```diff
@@ -125,7 +125,7 @@ static std::string attribute(const std::string& path, const std::string& value,
 {
   LM_T(LmtParse, ("Creating an attribute"));
   parseDataP->upcrs.attributeP = new ContextAttribute();
-  parseDataP->upcrs.attributeP->valueType = orion::ValueTypeNone;
+  parseDataP->upcrs.attributeP->valueType = orion::ValueTypeNotGiven;
```
Comment (kzangeli) [inline src/lib/jsonParse/jsonUpdateContextResponse.cpp]: Yeah, definitely.
All these assignments can be replaced by one line in the constructor (or, more lines if more constructors need it)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 4]
```diff
@@ -1843,9 +1813,13 @@ static void setPreviousValueMetadata(ContextElementResponse* notifyCerP)
...
 
-      case orion::ValueTypeNone:
+      case orion::ValueTypeNull:
         mdP = new Metadata(NGSI_MD_PREVIOUSVALUE, previousValueP->type, "");
-        mdP->valueType = orion::ValueTypeNone;
+        mdP->valueType = orion::ValueTypeNull;
+        break;
+
+      case orion::ValueTypeNotGiven:
+        LM_E(("Runtime Error (ValueTypeNotGiven value type)"));
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Like before, is it not better with English text here?
*(Especially as `ValueTypeNotGiven` is a bit cryptic)*

File: src/lib/ngsi/ContextAttribute.cpp [hunk 5]
```diff
@@ -828,10 +848,14 @@ std::string ContextAttribute::toJson
...
       out += JSON_VALUE_BOOL("value", boolValue);
     }
-    else if (valueType == orion::ValueTypeNone)
+    else if (valueType == orion::ValueTypeNull)
     {
       out += JSON_STR("value") + ":" + "null";
     }
+    else if (valueType == orion::ValueTypeNotGiven)
+    {
+      LM_E(("Runtime Error (ValueTypeNotGiven in compound value)"));
```
Comment (kzangeli) [inline src/lib/ngsi/ContextAttribute.cpp]: Not sure how we would end up here ... 
This is probably dead code, but to be safe, let's keep it.

The text in English would be:

    No value in compound value :-)

File: src/lib/orionTypes/OrionValueType.h [hunk 6]
```diff
@@ -34,16 +34,23 @@ namespace orion
 /* ****************************************************************************
 *
 * ValueType - 
+*
+* Note ValueTypeNotGiven is only used by NGSIv1 parsing logic. In NGSIv2 parsing,
+* attribute or metatadata value is implicit if ommitted (in particular, it
```
Comment (kzangeli) [inline src/lib/orionTypes/OrionValueType.h]: typo: omitted

File: src/lib/orionTypes/OrionValueType.h [hunk 7]
```diff
@@ -34,16 +34,23 @@ namespace orion
 /* ****************************************************************************
 *
 * ValueType - 
+*
+* Note ValueTypeNotGiven is only used by NGSIv1 parsing logic. In NGSIv2 parsing,
```
Comment (kzangeli) [inline src/lib/orionTypes/OrionValueType.h]: Note: ...
   or
Note that ...

File: src/lib/parse/CompoundValueNode.cpp [hunk 8]
```diff
@@ -883,6 +900,18 @@ std::string CompoundValueNode::toJson(bool isLastElement, bool comma)
       out = JSON_STR(key) + ":" + "null";
     }
   }
+  else if (valueType == orion::ValueTypeNotGiven)
+  {
+     LM_E(("Runtime Error (ValueTypeNotGiven (%s))", name.c_str()));
```
Comment (kzangeli) [inline src/lib/parse/CompoundValueNode.cpp]: indent

File: src/lib/common/macroSubstitute.cpp (hunk 1 above)
Comment (fgalan) [reply src/lib/common/macroSubstitute.cpp]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 4 above)
Comment (fgalan) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/orionTypes/OrionValueType.h (hunk 7 above)
Comment (fgalan) [reply src/lib/orionTypes/OrionValueType.h]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/parse/CompoundValueNode.cpp (hunk 8 above)
Comment (fgalan) [reply src/lib/parse/CompoundValueNode.cpp]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/orionTypes/OrionValueType.h (hunk 6 above)
Comment (fgalan) [reply src/lib/orionTypes/OrionValueType.h]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/ngsi/ContextAttribute.cpp (hunk 5 above)
Comment (fgalan) [reply src/lib/ngsi/ContextAttribute.cpp]: Fixed in e85892cdb28f2f510fd2b0a38cacbb43082f9613

File: src/lib/jsonParse/jsonAppendContextElementRequest.cpp (hunk 2 above)
Comment (fgalan) [reply src/lib/jsonParse/jsonAppendContextElementRequest.cpp]: Fixed in 707138e0b5ab6fc8c4642afa906290b48fd3f7ae

Not only the ContextAttribute constructor, but also Metatada constructor, with additional fixes in the metadata parsing logic suprisingly unnoticed until now. I'd suggest to review that commit .

File: src/lib/jsonParse/jsonUpdateContextResponse.cpp (hunk 3 above)
Comment (fgalan) [reply src/lib/jsonParse/jsonUpdateContextResponse.cpp]: Fixed in 707138e0b5ab6fc8c4642afa906290b48fd3f7ae

Not only the ContextAttribute constructor, but also Metatada constructor, with additional fixes in the metadata parsing logic suprisingly unnoticed until now. I'd suggest to review that commit .

File: src/lib/jsonParse/jsonRegisterContextRequest.cpp [hunk 9]
```diff
@@ -204,11 +204,6 @@ static std::string craMetadata(const std::string& path, const std::string& value
   LM_T(LmtParse, ("Creating a metadata"));
 
   reqDataP->rcr.attributeMetadataP = new Metadata();
-
-  reqDataP->rcr.attributeMetadataP->type  = "";
```
Comment (fgalan) [inline src/lib/jsonParse/jsonRegisterContextRequest.cpp]: This initialization was unneced. This is work for the constructor. Maybe it comes from ancient days in which constructor where much more limited...

Comment (fgalan) [reply src/lib/jsonParse/jsonRegisterContextRequest.cpp]: NTC (just informative)

**Comment by kzangeli** [comment]
LGTM

**Comment by fgalan** [comment]
We will wait until IoTP v5.0 releasing before merging this. However, the associated issued will be closed.

**Pull request merged**